### PR TITLE
refactor!: Template measurements on dimensions not parameters (EDM rework 4/4)

### LINKED ISSUE #350: Acts::Measurement infrastructure has unacceptable build overhead
After trimming a bunch of Eigen fat, the compile-time overhead of Acts' fitting unit tests and examples becomes bottlenecked by the Measurement infrastructure, and more precisely the "unique ordered sublists" variant that it provides. To quantify this, about 50% of template-related compilation time for these compilation units is spent processing that code alone.

---

Given an N-dimensional state space `(x1, x2, ..., xN)`, this mechanism provides an `std::variant` that goes over all possible tuples `std::variant<(x1), (x1, x2), (x1, x3), ..., (x1, xN), (x2), (x2, x3), ..., (x2, xN), (x1, x2), (x1, x2, x3), ..., (x1, x2, xN), ..., (x1, x2, ..., xN)>`.

As you can imagine, the combinatorics are not in our favor here, and the resulting `std::variant` has a lot of different alternatives. More precisely, since this variant is semantically equivalent to an `(std::optional<x1>, std::optional<x2>, ..., std::optional<xN>)` tuple with the additional constraint of excluding the "all optionals empty" null state, we can easily compute that the number of variants will be 2^N-1, which for the N=6 case that Acts is currently using means 63 different variants.

The overhead associated with generating this type and its basic methods is already quite significant in practice because libstdc++'s implementation of `std::variant` is not optimized for low compile-time overhead, but for high-quality error reporting. It sends the input types through various type traits for SFINAE and static assertions, and type traits themselves are not super cheap to instantiate.

Furthermore, the overhead is not restricted to the instantiation of that `std::variant` and its basic operations like the constructor. Every `std::visit` through that `std::variant` will have its associated code duplicated 63 times. So for example, templated code which has a RAM footprint of 10MB when instantiated once will get, in the worst case where the compiler does not manage to deduplicate anything, a RAM footprint of 630MB when fed through `std::visit`.

---

During my last conversation with @paulgessinger about this mechanism, I did not manage to understand why exactly this std::variant cannot be replaced with `std::tuple<std::optional<x1>, std::optional<x2>, ..., std::optional<xN>>`, or an optimized version thereof.

In general, I do not currently understand this mechanism well enough to see if we actually need it, let alone optimize it. All I can say, from my current perspective, is that getting rid of it, if at all possible, would likely be highly beneficial to Acts' compilation times and build RAM consumption.

@paulgessinger and @msmk0 have expressed interest in working on this. I wouldn't mind helping either, but I need to understand this better to work on it, so if I were to work on this someone first needs to sit down with me for a while until I fully get it.

See also #280 where Paul proposes an alternate design for both Measurement and MultiTrajectory.

### PR DESCRIPTION
**Pull request by msmk0**
BREAKING CHANGE: `Measurement`, `FittableMeasurement`, and `FittableVolumeMeasurement` are removed and replaced with new types.

This rewrites the measurement EDM types such that only the measurement size is fixed at compile time. Which parameters are stored is determined at runtime. To achieve this the following new types are introduced:

- `FixedSizeSubspace` and `VariableSizeSubspace`: These helper types encode a subspace that is part of a larger vector space. They store which axes of the full space are part of the vector and provide the functionality to transform vectors to/from the subspace. They are agnostic to how the full space and subspace are ultimately stored.
- `FixedSizeMeasurement` replaces `Measurement`: stores parameter values and covariance for a fixed size measurement using `FixedSizeSubspace` internally for the encoding. Contrary to the previous `Measurement`, `FixedSizeMeasuremt` stores only a source link and no reference object. This avoids potential problems that arise from the fact that different measurements (bound vs. free) require different reference objects. As a result of previous changes, the source link must provide a `GeometryIdentifier` that can identify both surfaces and volumes. In all use cases, the reference object is alreay known before the measurement is accessed, e.g. from the navigator in the Kalman filter. Thus, the reference object does not need to be stored and can be removed.
- `VariantMeasument` replaces `Fittable{Volume}Measurement`. Similar to the `FittableMeasurement` this stores all possible measurement combinations for a given parameter space in a `std::variant`. Based on `FixedSizeMeasurement` this only requires as many variant types as there are parameter space dimensions, i.e. 6 for bound or 8 for free parameters. This should remove the compile time combinatorial explosion that is described in #350.

This obsoletes the `ParameterSet` infrastructure. `ParameterSet` and all related functionality and types are removed. Track parameters are implemented directly using the appropriate vector and matrix types. 

This is the last PR in the EDM rework series.

Closes #350 .

### CODE REVIEW

**Comment by msmk0** [comment]
This is a preview for the final EDM rework PR(s). This is still work-in-progress but the PR description already contains all the expected changes. The design and implementation of the new types is complete (+- some bugfixes) but they are not yet setup to replace the existing types.

File: Core/include/Acts/Utilities/detail/Subspace.hpp [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+  /// Project a full vector into the subspace.
+  ///
+  /// @tparam fullspace_vector_t Veector type in the full space
+  /// @param full Vector in the full space
+  /// @return Subspace vector w/ just the configured axis components
+  ///
+  /// @note Always returns a column vector regardless of the input
+  template <typename fullspace_vector_t>
+  auto projectVector(const Eigen::MatrixBase<fullspace_vector_t>& full) const
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/detail/Subspace.hpp:147]: Ideally I think we should try to stick to the projector matrix and do the matrix mult externally. This duplicates that functionality right? If so, I'd argue for this to be removed.

File: Core/include/Acts/Utilities/detail/Subspace.hpp [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+
+  /// Expand a subspace vector into the full space.
+  ///
+  /// @tparam subspace_vector_t Subspace vector type
+  /// @param sub Subspace vector
+  /// @return Vector in the full space w/ zeros for non-subspace components
+  ///
+  /// @note Always returns a column vector regardless of the input
+  template <typename subspace_vector_t>
+  auto expandVector(const Eigen::MatrixBase<subspace_vector_t>& sub) const
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/detail/Subspace.hpp:166]: We don't ever do or need this do we?

File: Core/include/Acts/Utilities/detail/Subspace.hpp [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+    return expn;
+  }
+};
+
+/// Variable-size subspace representation.
+///
+/// @tparam kFullSize Size of the full vector space
+/// @tparam kMaxSize Maximum size of the subspace
+template <size_t kFullSize, size_t kMaxSize = kFullSize>
+class VariableSizeSubspace {
```
Comment (paulgessinger) [inline Core/include/Acts/Utilities/detail/Subspace.hpp:210]: Maybe I'm missing the point (still on my phone), but can you explain the use case of this?

**Comment by paulgessinger** [comment]
I feel like this is big step in the right direction. I'm unclear if some of the new types need to exist to support current functionality and would favor removing them if not. I can be persuaded here though. 

I would also suggest to keep the naming as before, i.e. Measurement rather than FixedSizeMeasurement, for consistency. There is not (and shouldn't really be) a dynamic sized measurement, so I don't think we need to differentiate here.

File: Core/include/Acts/Utilities/detail/Subspace.hpp (hunk 2 above)
Comment (asalzburger) [reply Core/include/Acts/Utilities/detail/Subspace.hpp:166]: I have a case where I use this, in fact. In the SmearingDigitization, if I remember correctly.

Comment (paulgessinger) [reply Core/include/Acts/Utilities/detail/Subspace.hpp:166]: I see. But that could also be done with the matrix multiplication externally. Do we expect that to be slower than this function?

**Comment by msmk0** [comment]
Superseded by #555 and #557.

**Pull request closed without merging**